### PR TITLE
Bryanv/categoricals

### LINKED ISSUE #516: Labels on Axis do not like ":" character ;-)
Good afternoon,

While settings label axis with character ":", labels was not displayed on the screen.
My need was to display time (format H:M:S) as xaxis label. Nothing change if I add  x_axis_type = "datetime".

Many thanks for Bokeh library and your works ...

Pegase92
# 

Code example that is running correctly : 

```
from bokeh.plotting import *
output_file("boxplot.html")
figure()
hold()
Etiquette=['A','B','C','D','E']

segment(Etiquette,[0,0,0,0,0,0,0,0,0,0,0],Etiquette,[-1,2,3,2,0],x_range=Etiquette,line_width=2, line_color="black",plot_width=1200,plot_height=500)  

rect(x=Etiquette,y=[-.5,1,1.5,1,0],width=0.5,height=[-1,2,3,2,0],fill_color='blue',plot_width=1200,plot_height=500,tools=TOOLS)

show()
```
# 

Code example that is not running correctly (xaxis labels are not displayed) : 

```
from bokeh.plotting import *
output_file("boxplot.html")
figure()
hold()
Etiquette=['00:00:00-00:00:01','00:00:01-00:00:02','00:00:02-00:00:03','00:00:03-00:00:04','00:00:04-00:00:05']

segment(Etiquette,[0,0,0,0,0,0,0,0,0,0,0],Etiquette,[-1,2,3,2,0],x_range=Etiquette,line_width=2, line_color="black",plot_width=1200,plot_height=500)  

rect(x=Etiquette,y=[-.5,1,1.5,1,0],width=0.5,height=[-1,2,3,2,0],fill_color='blue',plot_width=1200,plot_height=500,tools=TOOLS)

show()
```

### LINKED ISSUE #4369: Spaced hierarchical index
from a discussion on the mailing list:
https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/Lq5LX6yH-fU

---

Currently bokeh can have hierarchical indexes on plots, but they are slotted for each tick, see in this example:

``` python
import pandas as pd
import datetime
from bokeh.charts import Bar, output_notebook, show

output_notebook()

data = {
    'month': ['2016-01-01','2016-02-01', '2016-01-01'],
    'work_location': ['site_a', 'site_a', 'site_b'],
    'nps': [50.0, 33.3, -25.0]
}

df = pd.DataFrame(data)

p = Bar(df, values='nps', label=['work_location', 'month'])
show(p)
```

Produces the following plot:
plot_examples

On the left is the current, on the right the axis is parsed visually so it is easier/quicker to interpret and see the groups.

Here is an additional example:
http://news.infragistics.com/cfs-filesystemfile.ashx/__key/CommunityServer.Discussions.Components.Files/265/4718.Capture.PNG

Again, just floating the idea as I think it lends visual power to the graph in making the content of the axis (when it is hierarchical like this) easier to parse than the current look.

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #4369
issue: fixes #516

Allows for nested categoricals, upt to three levels, much easier to produce e.g. bar charts like this:

screen shot 2017-08-03 at 11 03 50

Some things were removed, at least for now:

* `offset` 

This was only for MPL compat which is being dropped, so I have no qualms removing it entirely. Some things were changed (these now accept synthetic coordinates, which is much simpler/more straighforward)

* `bounds`
* `min_interval` 
* `max_interval`

Planned work:

* [x] expand `factors` property to accept nested categories in addition to simple lists of factors
* [x] update `synthetic` to properly compute synthetic coords for hierarchical categories
* [x] add support for display padding between categories 
* [x] update axis to be able to display ranges for multi-level categories
* [x] make CategoricalColorMapper support nested categories
* [x] fix up JS API
* [x] computed expressions
* [x] sugar for dodging, jitter, and stacking things
* [x] add tests
* [ ] add examples
* [ ] add docs

ping @bokeh/dev

### CODE REVIEW

**Comment by bryevdv** [comment]
@mattpap JS tests are failing due to one thing, I had to add this to `set_data`:
```
    # if we have any coordinates that are categorical, convert them to
    # synthetic coords here
    xr = @renderer.plot_view.frame.x_ranges[@model.x_range_name]
    yr = @renderer.plot_view.frame.y_ranges[@model.y_range_name]
    for [xname, yname] in @model._coords
      xname = "_#{xname}"
      yname = "_#{yname}"
      if xr.v_synthetic?
        @[xname] = xr.v_synthetic(@[xname])
      if yr.v_synthetic?
        @[yname] = yr.v_synthetic(@[yname])
``` 
but `x_ranges` etc are undefined, and things fall over. Any suggestions?

**Comment by bryevdv** [comment]
Added 
```
if @renderer.plot_view.frame.x_ranges? 
``` 
check as a temp workaround for now

**Comment by mattpap** [comment]
> but x_ranges etc are undefined, and things fall over. Any suggestions?

Are offending tests using fully initialized models? Looking at the code, if models are properly initialized, this shouldn't happen. Obviously it can be a dozen of other things. If I knew what tests are failing, I could say more.

**Comment by bryevdv** [comment]
> Are offending tests using fully initialized models?

Sorry, it's the defaults test

**Comment by canavandl** [comment]
I just had an idea of something to check before merging:

In the case of a VBar with the "x" values in categorical space, presumably a numerical width value should still work.

**Comment by bryevdv** [comment]
Early example (ping @pzwang)
```
from bokeh.io import output_file, show
from bokeh.models import FactorRange
from bokeh.plotting import figure
from bokeh.sampledata.autompg import autompg as df

df.origin = df.origin.astype(str)
df.cyl = df.cyl.astype(str)
group = df.groupby(('origin', 'cyl'))
stats = group.mean()

p = figure(x_range=FactorRange(*g.groups.keys()), y_range=(0,35))
p.vbar(x=stats.index, top=stats.mpg, width=0.9, color=color)

output_file("foo.html")

show(p)
```

screen shot 2017-07-23 at 15 11 00

@canavandl it does, when I've got a bit more done I will write up the tradeoffs. One of them is that factors have to be strings. I don't see an alternative, otherwise numbers become ambiguous "is this a factor? or an offset?"

**Comment by canavandl** [comment]
> @canavandl it does, when I've got a bit more done I will write up the tradeoffs. One of them is that factors have to be strings. I don't see an alternative, otherwise numbers become ambiguous "is this a factor? or an offset?"

Great! I was just had the idea while making an unrelated example and wanted to ensure that edge case is handled.

**Comment by bryevdv** [comment]
> Great! I was just had the idea while making an unrelated example and wanted to ensure that edge case is handled.

This is actually other change. Before, every category offset was a percentage, with 0 being the prev cat, and 1 being the next cat, and 0.5 being the cat in question. Now its not like that. The offset is simply a +/- added to a factor's synthetic coordinate. It's mostly all better and simpler but there is one trade off. If you have a top level category with three sub categories and you wanted a big bar 90% wide of the full top-level cat, you have to know to make the width 3*0.9 = 2.7 sucks a bit but is not a common case and also not so terrible.

**Comment by mattpap** [comment]
> Sorry, it's the defaults test.

That makes sense, because those tests work with insufficiently configured models. However, this piece of code shouldn't be triggered by those tests.

**Comment by bryevdv** [comment]
@mattpap I'd like to extend this to handle more levels of nesting (at least one or two more) It would really make things simple if there was a Map object that could use list values as keys. Do you know of one we could add and use that is lightweight?

**Comment by mattpap** [comment]
@bryevdv, what operations on that map you need to perform? You could use ES6 Map, but that is very annoying to iterate over in ES5.

**Comment by bryevdv** [comment]
@mattpap I just want to be able to do e.g.
```
mapping[ ["foo", "bar"] ] = 10
``` 
and have it be robust (i.e. not "work" by converting the list to a string and hashing that)

**Comment by mattpap** [comment]
@bryevdv, so use ES6 Map, e.g.
```
map = new Map()
map.set([1, 2, 3], ["a", "b", "c"])
map.get([1, 2, 3])
map.has([1, 2, 3])
```
If you need iteration then there is `forEach`:
```
map.forEach((key, value) -> ...)
```

**Comment by bryevdv** [comment]
@mattpap the `.has` method seems to be failing, at least with lists:

screen shot 2017-07-25 at 17 00 06

**Comment by mattpap** [comment]
@bryevdv, right, it uses `===` (with a special case for `NaN`) semantics for comparison, and `===` is useless for pretty much anything than primitives.

I will implement something more reasonable, as I need a proper map for my own work anyway.

File: bokehjs/src/coffee/models/ranges/factor_range.coffee [hunk 1]
```diff
@@ -1,62 +1,167 @@
...
-      factors: [ p.Array,  [] ]
-      bounds:  [ p.Any        ] # TODO (bev)
-      min_interval: [ p.Any ]
-      max_interval: [ p.Any ]
-    }
+    factors:             [ p.Array,        []        ]
+    factor_padding:      [ p.Number,       0         ]
+    subgroup_padding:    [ p.Number,       0.8       ]
+    group_padding:       [ p.Number,       1.4       ]
+    factor_padding:      [ p.Number,       0         ]
```
Comment (canavandl) [inline bokehjs/src/coffee/models/ranges/factor_range.coffee]: factor_padding is here twice

**Comment by bryevdv** [comment]
@mattpap @canavandl @clairetang6 I still need to add docs and examples, but this is basically ready for at least a preliminary technical review.

File: bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee [hunk 2]
```diff
@@ -1,22 +1,56 @@
...
-import {ColorMapper} from "./color_mapper"
+_equals = (a, b) ->
+  if a.length != b.length
+    return false
+  for i in [0...a.length]
+    if a[i] != b[i]
+      return false
+  return true
+
+_find = (a, b) ->
```
Comment (mattpap) [inline bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee]: `core/util/array` has `findIndex()`.

Comment (bryevdv) [inline bokehjs/src/coffee/models/mappers/categorical_color_mapper.coffee]: Thanks @mattpap I have updated and added full tests for the categorical colormapper as well

**Comment by bryevdv** [comment]
@bokeh/dev this is marked ready, the only additional work planned are docs, examples and (possibly) a tiny convenience to accept pandas groupby's directly for factors to `figure` (very simple to do)

**Comment by bryevdv** [comment]
@mattpap @canavandl ideally I'd like to get this merged by Fri/Sat and in a new dev build by Monday so folks can start banging on it

**Comment by bryevdv** [comment]
I added a an optional `range` prop to `Jitter` so that the transform can work with categorical data:
```
import pandas as pd

from bokeh.io import output_file, show
from bokeh.models import ColumnDataSource, Jitter
from bokeh.plotting import figure

output_file("stats.html")

DAYS = ['Sun', 'Sat', 'Fri', 'Thu', 'Wed', 'Tue', 'Mon']

data = pd.read_csv("commits.txt", parse_dates=True, header=None, names=['day', 'datetime'], index_col='datetime')
data = data.tz_localize('GMT').tz_convert('US/Central')
data['time'] = data.index.time

source = ColumnDataSource(data)

p = figure(plot_width=800, plot_height=300, y_range=DAYS, x_axis_type='datetime', title="Bokeh Commits (US/Central) ")
p.circle(x='time', y={'field': 'day', 'transform': Jitter(width=0.7, range=p.y_range)},  source=source, alpha=0.5)

p.xaxis[0].formatter.days = ['%Hh']
p.ygrid.grid_line_color = None

show(p)
```
screen shot 2017-08-03 at 14 18 16

**Comment by jbednar** [comment]
Nice!

**Comment by bryevdv** [comment]
OK, last small feature, can pass pandas groups directly to `figure` to set ranges:
```
group = df.groupby(('origin', 'cyl'))

p = figure(plot_width=1100, x_range=group, tools="hover,pan,wheel_zoom,reset")
```

**Comment by bryevdv** [comment]
ok, really, truly last feature, CDS can now be inited with Pandas groupbys, which will create columns  from `group.describe`:
```
group = df.groupby(('origin', 'cyl'))
stats = group.mean()

source = ColumnDataSource(group)
cmap = CategoricalColorMapper(palette=Spectral5, factors=sorted(df.cyl.unique()), start=1)

p = figure(plot_width=1100, x_range=group, tools="hover,pan,wheel_zoom,reset")
p.vbar(x='origin_cyl', top='mpg_mean', width=1, line_color="white", fill_color={'field': 'origin_cyl', 'transform': cmap}, source=source, legend="origin_cyl")
```

All this goes a huge way to making `bokeh.plotting` much more friendly to categorical data and pandas as well. I will work on docs/examples tomorrow

**Comment by bryevdv** [comment]
There was a small issue with cat color mapper. I had a bad test but picked it up writing examples. I will push the fix with some docs later today

**Comment by bryevdv** [comment]
OK, I made one more somewhat substantial addition. I added support for computed expressions in https://github.com/bokeh/bokeh/pull/6654/commits/cc5ba864220070c77af463594a6ad40be5fee369 This has been mentioned on and off for a long time, and it is necessary to round out a good usr experience for categoricals. Fortunately in the current state, adding it was fairly straightforward and required little code change and no existing test changes. With this addition, it is possible to create a `Stack` expression that permits code like this:

```
source = ColumnDataSource(data={'fruits' : fruits,
                                '2016'   : [5, 3, 4, 2, 4, 6],
                                '2017'   : [3, 2, 4, 4, 5, 3]})

p = figure(x_range=fruits, plot_height=250,
    toolbar_location=None, title="Fruit Counts by Year")

p.vbar(x='fruits', top='2016', width=0.9, color="#718dbf",
       source=source, legend=value('2016'))
p.vbar(x='fruits', bottom=stack('2016'), top=stack('2016', '2017'),
       width=0.9, color="#e84d60", source=source, legend=value('2017'))
```
which easily yields the ever asked-for but otherwise far too-difficult stacked bar chart:

screen shot 2017-08-04 at 21 58 51

I've added this example to the howto notebook, and will start the user guide chapter based off the notebook tomorrow. 

@mattpap  https://github.com/bokeh/bokeh/pull/6654/commits/cc5ba864220070c77af463594a6ad40be5fee369 has good tests but please do review specifically.

**Comment by bryevdv** [comment]
@bokeh/dev plan to merge this later today and make a dev build and put out a general call for help exercising it on the mailing list, unless there is objection

**Comment by bryevdv** [comment]
The only notable last change was https://github.com/bokeh/bokeh/pull/6654/commits/cb5f3ae712f4916f008057b19050153c11637903 which slightly simplified CDS init from data frame to handle cases where there is a `None` index name more gracefully (test added)

**Pull request merged**